src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -394,5 +444,51 @@ public TestAccessor(CompletionHandler completionHandler)
             public CompletionListCache GetCache()
                 => _completionHandler._completionListCache;
         }
+
+        private class CommitCharacterArrayComparer : IEqualityComparer<ImmutableArray<CharacterSetModificationRule>>
+        {
+            public static readonly CommitCharacterArrayComparer Instance = new();
+
+            private CommitCharacterArrayComparer()
+            {
+            }
+
+            public bool Equals([AllowNull] ImmutableArray<CharacterSetModificationRule> x, [AllowNull] ImmutableArray<CharacterSetModificationRule> y)
+            {
+                for (var i = 0; i < x.Length; i++)
+                {
+                    var xKind = x[i].Kind;
+                    var yKind = y[i].Kind;
+                    if (xKind != yKind)
+                    {
+                        return false;
+                    }
+
+                    var charactersX = x[i].Characters;
```

Comment:
super nit: for naming consistency with `x/yKind`, maybe `x/yCharacters` might match better here?